source/postprocess/visualization/maximum_horizontal_compressive_stress.cc

```diff
@@ -47,134 +47,161 @@ namespace aspect
+                  {
+                    compressive_stress -= elastic_additional_out->deviatoric_stress[q];
+                  }
+                else
+                  {
+                    compressive_stress -= 2. * eta * deviatoric_strain_rate;
+                  }
+              }
+            else
+              {
+                compressive_stress -= 2. * eta * deviatoric_strain_rate;
+              }
+
+            pack_tensor(compressive_stress, stress_components[q]);
+          }
+
+        // --- STEP 2: average stress components if requested (use existing visualization setting)
+        const auto &viz = this->get_postprocess_manager().template get_matching_active_plugin<Postprocess::Visualization<dim>>();
+        if (!viz.output_pointwise_stress_and_strain())
+          average_quantities(stress_components);
+
+        // --- STEP 3: from averaged stress components compute final direction and magnitude
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            SymmetricTensor<2,dim> avg_compressive_stress;
+            unpack_tensor(stress_components[q], avg_compressive_stress);
+
             const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector (in.position[q]);
-            const Tensor<1,dim> vertical_direction = gravity/gravity.norm();
-            std::array<Tensor<1,dim>,dim-1 > orthogonal_directions
-              = Utilities::orthogonal_vectors(vertical_direction);
+            const Tensor<1,dim> vertical_direction = gravity / gravity.norm();
+            std::array<Tensor<1,dim>,dim-1> orthogonal_directions = Utilities::orthogonal_vectors(vertical_direction);
             for (unsigned int i=0; i<orthogonal_directions.size(); ++i)
               orthogonal_directions[i] /= orthogonal_directions[i].norm();
 
             Tensor<1,dim> maximum_horizontal_compressive_stress;
             switch (dim)
               {
-                // in 2d, there is only one horizontal direction, and
-                // we have already computed it above. give it the
-                // length of the compressive_stress (now taking into account the
-                // pressure) in this direction
                 case 2:
-                {
                   maximum_horizontal_compressive_stress = orthogonal_directions[0] *
                                                           (orthogonal_directions[0] *
-                                                           ((compressive_stress
-                                                             -
-                                                             in.pressure[q] * unit_symmetric_tensor<dim>()) *
+                                                           (avg_compressive_stress *
                                                             orthogonal_directions[0]));
                   break;
-                }
 
-                // in 3d, use the formulas discussed in the
-                // documentation of the plugin below
```

Comment:
Also keep this comment. It is not super important, but informative on where the formulas come from.